File: test/files/snake_PP.vb
' VB.NET with Elan 2.0.0-beta4

' Use the w,a,s,d keys to change snake's direction

Const width = 40

Const height = 30

Sub main()
  Dim bg = New BlockGraphics() ' variable definition
  Dim head = 621 ' variable definition
  Dim snake = {head - 1, head} ' variable definition
  Dim currentDir = "d" ' variable definition
  Dim gameOn = True ' variable definition
  Dim apple = -1 ' variable definition
  While gameOn
    While (apple = -1) Or snake.contains(apple)
      apple = randint(0, width*height) ' assignment
    End While
    updateDisplay(bg, snake, apple) ' procedure call
    Dim key = getKey() ' variable definition
    If Not key.equals("") And "wasd".contains(key) Then
      currentDir = key ' assignment
    End If
    head = getAdjacentBlock(head, currentDir) ' assignment
    If (head = -1) Or snake.contains(head) Then
      gameOn = False ' assignment
    Else
      snake.append(head) ' procedure call
    End If
    If head.equals(apple) Then
      apple = -1 ' assignment
    Else
      snake.removeAt(0) ' procedure call
    End If
    sleep_ms(150) ' procedure call
  End While
  Console.WriteLine($"Game Over! Score: {snake.length() - 1}") ' print statement
End Sub

Sub updateDisplay(bg As BlockGraphics, snake As List(Of Integer), apple As Integer) ' procedure
  bg.colourAll(white) ' procedure call
  For Each bl In snake
    bg.putBlockNo(bl, green) ' procedure call
  Next bl
  bg.putBlockNo(apple, red) ' procedure call
  displayBlockGraphics(bg) ' procedure call
End Sub

Function getAdjacentBlock(bl As Integer, dir As String) As Integer
  Dim adj = -1 ' variable definition
  Dim newCol = bl Mod width ' variable definition
  Dim newRow = divAsInt(bl, width) ' variable definition
  If dir.equals("w") Then
    newRow = newRow - 1 ' assignment
  ElseIf dir.equals("a") Then
    newCol = newCol - 1 ' assignment
  ElseIf dir.equals("s") Then
    newRow = newRow + 1 ' assignment
  ElseIf dir.equals("d") Then
    newCol = newCol + 1 ' assignment
  End If
  If (newCol >= 0) And (newCol < width) And (newRow >= 0) And (newRow < height) Then
    adj = newRow*width + newCol ' assignment
  End If
  Return adj
End Function

<TestClass Class Test_getAdjacentBlock
 <TestMethod> Sub test_getAdjacentBlock()
  Dim bl = 617 ' variable definition
  Assert.AreEqual(577, getAdjacentBlock(bl, "w"))
  Assert.AreEqual(616, getAdjacentBlock(bl, "a"))
  Assert.AreEqual(657, getAdjacentBlock(bl, "s"))
  Assert.AreEqual(618, getAdjacentBlock(bl, "d"))
  ' boundary
  Assert.AreEqual(-1, getAdjacentBlock(20, "w"))
  Assert.AreEqual(-1, getAdjacentBlock(40, "a"))
  Assert.AreEqual(-1, getAdjacentBlock(1180, "s"))
  Assert.AreEqual(-1, getAdjacentBlock(79, "d"))
 End Sub
End Class

File: test/files/snake_PP.cs
// C# with Elan 2.0.0-beta5

// Use the w,a,s,d keys to change snake's direction

const int width = 40;

const int height = 30;

static void main() {
  var bg = new BlockGraphics();
  var head = 621;
  var snake = new [] {head - 1, head};
  var currentDir = "d";
  var gameOn = true;
  var apple = -1;
  while (gameOn) {
    while ((apple == -1) || snake.contains(apple)) {
      apple = randint(0, width*height); // assignment
    } // end while
    updateDisplay(bg, snake, apple); // procedure call
    var key = getKey();
    if (!key.equals("") && "wasd".contains(key)) {
      currentDir = key; // assignment
    } // end if
    head = getAdjacentBlock(head, currentDir); // assignment
    if ((head == -1) || snake.contains(head)) {
      gameOn = false; // assignment
    } else {
      snake.append(head); // procedure call
    } // end if
    if (head.equals(apple)) {
      apple = -1; // assignment
    } else {
      snake.removeAt(0); // procedure call
    } // end if
    sleep_ms(150); // procedure call
  } // end while
  Console.WriteLine($"Game Over! Score: {snake.length() - 1}"); // print statement
} // end main

static void updateDisplay(BlockGraphics bg, List<int> snake, int apple) { // procedure
  bg.colourAll(white); // procedure call
  foreach (var bl in snake) {
    bg.putBlockNo(bl, green); // procedure call
  } // end foreach
  bg.putBlockNo(apple, red); // procedure call
  displayBlockGraphics(bg); // procedure call
} // end procedure

static int getAdjacentBlock(int bl, string dir) { // function
  var adj = -1;
  var newCol = bl % width;
  var newRow = divAsInt(bl, width);
  if (dir.equals("w")) {
    newRow = newRow - 1; // assignment
  } else if (dir.equals("a")) {
    newCol = newCol - 1; // assignment
  } else if (dir.equals("s")) {
    newRow = newRow + 1; // assignment
  } else if (dir.equals("d")) {
    newCol = newCol + 1; // assignment
  } // end if
  if ((newCol >= 0) && (newCol < width) && (newRow >= 0) && (newRow < height)) {
    adj = newRow*width + newCol; // assignment
  } // end if
  return adj;
} // end function

[TestClass] class Test_getAdjacentBlock
[TestMethod] static void test_getAdjacentBlock() {
  var bl = 617;
  Assert.AreEqual(577, getAdjacentBlock(bl, "w"));
  Assert.AreEqual(616, getAdjacentBlock(bl, "a"));
  Assert.AreEqual(657, getAdjacentBlock(bl, "s"));
  Assert.AreEqual(618, getAdjacentBlock(bl, "d"));
  // boundary
  Assert.AreEqual(-1, getAdjacentBlock(20, "w"));
  Assert.AreEqual(-1, getAdjacentBlock(40, "a"));
  Assert.AreEqual(-1, getAdjacentBlock(1180, "s"));
  Assert.AreEqual(-1, getAdjacentBlock(79, "d"));
}} // end test
